# Pokemon Web Scraping | Viraj Shah

In [1]:
import numpy as np
import pandas as pd
import requests
import os 
from bs4 import BeautifulSoup
import re
from IPython.display import display, HTML
from urllib.request import urlopen
from IPython.display import Image

## Exercise #1  - Scraping the pokedex

### Step 1.1 : Use BeautifulSoup to extract all the table rows as a list

In [2]:
url=requests.get("https://pokemondb.net/pokedex/all")
# Parse the html content
soup = BeautifulSoup(url.content, 'html.parser')
# print(soup.prettify()) # print the parsed data of html
y = soup.find_all(name = "table", class_="data-table block-wide")
# print (y)
all_list = []
for tr in y[0].find_all("tr"):
    # remove any new lines and extra spaces from left and right
    all_list.append(tr.get_text().replace('\n', ' ').strip())
print(all_list[1:10])

['001 BulbasaurGrass Poison 318 45 49 49 65 65 45', '002 IvysaurGrass Poison 405 60 62 63 80 80 60', '003 VenusaurGrass Poison 525 80 82 83 100 100 80', '003 Venusaur Mega VenusaurGrass Poison 625 80 100 123 122 120 80', '004 CharmanderFire  309 39 52 43 60 50 65', '005 CharmeleonFire  405 58 64 58 80 65 80', '006 CharizardFire Flying 534 78 84 78 109 85 100', '006 Charizard Mega Charizard XFire Dragon 634 78 130 111 130 85 100', '006 Charizard Mega Charizard YFire Flying 634 78 104 78 159 115 100']


### How many rows are there (including the header row)?

In [3]:
print("Total number of rows (including the header row) : " + (str(len(all_list))))

Total number of rows (including the header row) : 1029


### Step 1.2 : Save the first row of the table (bulbasaur) as a variable. Using .find(), and .find_all(), extract and print the following contents from that row:

### The name of the pokemon

In [4]:
print("The Pokemon's name is : " + soup.find(name="td", class_="cell-name").text)

The Pokemon's name is : Bulbasaur


### The url to the pokemon's page

In [5]:
print("The url to Pokemon's page : " + soup.find_all(name="a", class_="ent-name")[0].get("href"))

The url to Pokemon's page : /pokedex/bulbasaur


### The type or types (as a string separated by spaces, e.g. "Fire Flying")

In [6]:
types = soup.find_all(name="td", class_="cell-icon")[0].get_text()
print("The types are : " + types)

The types are : Grass Poison


### The total points

In [7]:
print("The total points : " + soup.find_all(name="td", class_="cell-total")[0].text)

The total points : 318


### In a single list (via appending):  ID Number, HP, Attack, Defense, Sp. Atk, Sp. Def, Speed

In [8]:
global new_list 
new_list = []
cellnum = soup.find_all(name="td", class_="cell-num")
for i in cellnum:
    new_list.append(i.get_text())
print(new_list[:7])

['001', '45', '49', '49', '65', '65', '45']


### Step 1.3 : Generalize step 1.2. Define a function that takes in a row of the pokedex table and returns it as a DataFrame with a single row. Create a single DataFrame by appending these rows.

In [9]:
def pokemon():
    global pokemon_ids,pokemon_names,pokemon_types,pokemon_total,pokemon_hp,pokemon_attack,pokemon_defense,pokemon_spatt,pokemon_spdef,pokemon_speed
    pokemon_ids = []
    pokemon_names = []
    pokemon_types = []
    pokemon_total = []
    pokemon_hp = []
    pokemon_attack = []
    pokemon_defense = []
    pokemon_spatt = []
    pokemon_spdef = []
    pokemon_speed = []
    
    ids = soup.find_all(name = "span", class_ = "infocard-cell-data")
    [pokemon_ids.append(x.get_text()) for x in ids]
    #print(pokemon_ids[:10])
    
    name = soup.find_all(name = "td", class_ = "cell-name")
    [pokemon_names.append(x.get_text()) for x in name]
    #print(pokemon_names[:10])
    
    types = soup.find_all(name = "td", class_="cell-icon")
    [pokemon_types.append(x.get_text()) for x in types]
    #print(pokemon_types[:10])
    
    total = soup.find_all(name = "td", class_ = "cell-total")
    [pokemon_total.append(x.get_text()) for x in total]
    #print(pokemon_total[:10])
    
    hp = soup.find_all(name = "td", class_ = "cell-num")
    [pokemon_hp.append(new_list[x]) for x in range(1,len(new_list),7)]
    #print(pokemon_hp[:10])
    
    attack = soup.find_all(name = "td", class_ = "cell-num")
    [pokemon_attack.append(new_list[x]) for x in range(2, len(new_list), 7)]
    #print(pokemon_attack[:10])
    
    defense = soup.find_all(name = "td", class_ = "cell-num")
    [pokemon_defense.append(new_list[x]) for x in range(3, len(new_list), 7)]
    
    defense = soup.find_all(name = "td", class_ = "cell-num")
    [pokemon_spatt.append(new_list[x]) for x in range(4, len(new_list), 7)]
    
    defense = soup.find_all(name = "td", class_ = "cell-num")
    [pokemon_spdef.append(new_list[x]) for x in range(5, len(new_list), 7)]
    
    defense = soup.find_all(name = "td", class_ = "cell-num")
    [pokemon_speed.append(new_list[x]) for x in range(6, len(new_list), 7)]
    
    pokemon_data = [pokemon_ids, pokemon_names, pokemon_types, pokemon_total, pokemon_hp, pokemon_attack, pokemon_defense,
                    pokemon_spatt, pokemon_spdef, pokemon_speed]
    
    df = pd.DataFrame(pokemon_data)
    global df1
    df1 = df.T
    d = df1.head()
    p = df1.tail()
    display(HTML(d.to_html(index=False)))
    display(HTML(p.to_html(index=False)))

In [10]:
pokemon()

## Exercise #2 - Cleaning the Pokedex

### Step 2.1 : Add column names to the DataFrame. Convert strings to numeric where appropriate. Make the ID number the first column in the DataFrame if it is not already.

In [11]:
# converting to int from string
pokemon_ids = [int (i) for i in pokemon_ids]
pokemon_hp = [int (i) for i in pokemon_hp]
pokemon_attack = [int (i) for i in pokemon_attack]
pokemon_defense = [int (i) for i in pokemon_defense]
pokemon_spatt = [int (i) for i in pokemon_spatt]
pokemon_spdef = [int (i) for i in pokemon_spdef]
pokemon_speed = [int (i) for i in pokemon_speed]

# providing column names
pokemon_dict = {"Id" : pokemon_ids, "Name" : pokemon_names, "Type" : pokemon_types, "Total" : pokemon_total,
                "HP" : pokemon_hp, "Att." : pokemon_attack, "Def." : pokemon_defense, "Sp. Att." : pokemon_spatt,
                "Sp. Def." : pokemon_spdef, "Speed" : pokemon_speed}
df1 = pd.DataFrame(pokemon_dict)
u = df1.head()
o = df1.tail()
display(HTML(u.to_html(index=False)))
display(HTML(o.to_html(index=False)))

### Step 2.2 : Notice that the pokemon types are not mutually exclusive. (Pokemon may have more than one type.) Create 18 dummy variables for each type of pokemon. 

In [12]:
# creating 18 dummy variables for each type of pokemon
dum = df1['Type'].str.get_dummies(' ')
dum

Bug  Dark  Dragon  Electric  Fairy  Fighting  Fire  Flying  Ghost  \
0       0     0       0         0      0         0     0       0      0   
1       0     0       0         0      0         0     0       0      0   
2       0     0       0         0      0         0     0       0      0   
3       0     0       0         0      0         0     0       0      0   
4       0     0       0         0      0         0     1       0      0   
...   ...   ...     ...       ...    ...       ...   ...     ...    ...   
1023    0     0       0         0      1         0     0       0      0   
1024    0     0       0         0      0         1     0       0      0   
1025    0     0       0         0      0         1     0       0      0   
1026    0     0       1         0      0         0     0       0      0   
1027    0     0       1         0      0         0     0       0      0   

      Grass  Ground  Ice  Normal  Poison  Psychic  Rock  Steel  Water  
0         1       0    0       0       1        0     0      0      0  
1         1       0    0       0       1        0     0      0      0  
2         1       0    0       0       1        0     0      0      0  
3         1       0    0       0       1        0     0      0      0  
4         0       0    0       0       0        0     0      0      0  
...     ...     ...  ...     ...     ...      ...   ...    ...    ...  
1023      0       0    0       0       0        0     0      0      0  
1024      0       0    0       0       0        0     0      1      0  
1025      0       0    0       0       0        0     0      0      0  
1026      0       0    0       0       1        0     0      0      0  
1027      0       0    0       0       1        0     0      0      0  

[1028 rows x 18 columns]

### Step 2.3 : Remove duplicate values of pokemon based on the URL. Keep the first observation in the case of a duplicate. 

In [13]:
pokemon_url = []
z = soup.find_all(name="a", class_="ent-name")
for x in z:
    pokemon_url.append(x.get("href"))
df1['URL'] = pokemon_url

In [14]:
df1 = df1.drop_duplicates(subset = 'URL', keep ='first', inplace = False)

In [15]:
df1

Id                      Name            Type Total   HP  Att.  Def.  \
0       1                 Bulbasaur    Grass Poison   318   45    49    49   
1       2                   Ivysaur    Grass Poison   405   60    62    63   
2       3                  Venusaur    Grass Poison   525   80    82    83   
4       4                Charmander           Fire    309   39    52    43   
5       5                Charmeleon           Fire    405   58    64    58   
...   ...                       ...             ...   ...  ...   ...   ...   
1020  886                  Drakloak    Dragon Ghost   410   68    80    50   
1021  887                 Dragapult    Dragon Ghost   600   88   120    75   
1022  888      Zacian Crowned Sword     Fairy Steel   720   92   170   115   
1024  889  Zamazenta Crowned Shield  Fighting Steel   720   92   130   145   
1026  890                 Eternatus   Poison Dragon   690  140    85    95   

      Sp. Att.  Sp. Def.  Speed                  URL  
0           65        65     45   /pokedex/bulbasaur  
1           80        80     60     /pokedex/ivysaur  
2          100       100     80    /pokedex/venusaur  
4           60        50     65  /pokedex/charmander  
5           80        65     80  /pokedex/charmeleon  
...        ...       ...    ...                  ...  
1020        60        50    102    /pokedex/drakloak  
1021       100        75    142   /pokedex/dragapult  
1022        80       115    148      /pokedex/zacian  
1024        80       145    128   /pokedex/zamazenta  
1026       145        95    130   /pokedex/eternatus  

[890 rows x 11 columns]

### Print the number of rows in the deduplicated dataset.

In [16]:
print("The number of rows in the deduplicated dataset : " + str(len(df1)))

The number of rows in the deduplicated dataset : 890


### Step 2.4 : Add a dummy variable to the DataFrame called "sample" that tags every 4th pokemon to be included in the sample

In [17]:
df1.loc[df1.Id % 4 == 0 , 'sample'] = '1' 
df1.loc[df1.Id % 4 > 0, 'sample'] = '0'

In [18]:
df1

Id                      Name            Type Total   HP  Att.  Def.  \
0       1                 Bulbasaur    Grass Poison   318   45    49    49   
1       2                   Ivysaur    Grass Poison   405   60    62    63   
2       3                  Venusaur    Grass Poison   525   80    82    83   
4       4                Charmander           Fire    309   39    52    43   
5       5                Charmeleon           Fire    405   58    64    58   
...   ...                       ...             ...   ...  ...   ...   ...   
1020  886                  Drakloak    Dragon Ghost   410   68    80    50   
1021  887                 Dragapult    Dragon Ghost   600   88   120    75   
1022  888      Zacian Crowned Sword     Fairy Steel   720   92   170   115   
1024  889  Zamazenta Crowned Shield  Fighting Steel   720   92   130   145   
1026  890                 Eternatus   Poison Dragon   690  140    85    95   

      Sp. Att.  Sp. Def.  Speed                  URL sample  
0           65        65     45   /pokedex/bulbasaur      0  
1           80        80     60     /pokedex/ivysaur      0  
2          100       100     80    /pokedex/venusaur      0  
4           60        50     65  /pokedex/charmander      1  
5           80        65     80  /pokedex/charmeleon      0  
...        ...       ...    ...                  ...    ...  
1020        60        50    102    /pokedex/drakloak      0  
1021       100        75    142   /pokedex/dragapult      0  
1022        80       115    148      /pokedex/zacian      1  
1024        80       145    128   /pokedex/zamazenta      0  
1026       145        95    130   /pokedex/eternatus      0  

[890 rows x 12 columns]

In [19]:
df2 = df1[df1['sample']=='1']
df2

Id                  Name             Type Total  HP  Att.  Def.  \
4       4            Charmander            Fire    309  39    52    43   
10      8             Wartortle           Water    405  59    63    80   
15     12            Butterfree       Bug Flying   395  60    45    50   
20     16                Pidgey    Normal Flying   251  40    45    40   
26     20              Raticate          Normal    413  55    81    60   
...   ...                   ...              ...   ...  ..   ...   ...   
1003  872                  Snom          Ice Bug   185  30    25    35   
1008  876         Indeedee Male   Psychic Normal   475  60    65    55   
1014  880             Dracozolt  Electric Dragon   505  90   100    90   
1018  884             Duraludon     Steel Dragon   535  70    95   115   
1022  888  Zacian Crowned Sword      Fairy Steel   720  92   170   115   

      Sp. Att.  Sp. Def.  Speed                  URL sample  
4           60        50     65  /pokedex/charmander      1  
10          65        80     58   /pokedex/wartortle      1  
15          90        80     70  /pokedex/butterfree      1  
20          35        35     56      /pokedex/pidgey      1  
26          50        70     97    /pokedex/raticate      1  
...        ...       ...    ...                  ...    ...  
1003        45        30     20        /pokedex/snom      1  
1008       105        95     95    /pokedex/indeedee      1  
1014        80        70     75   /pokedex/dracozolt      1  
1018       120        50     85   /pokedex/duraludon      1  
1022        80       115    148      /pokedex/zacian      1  

[222 rows x 12 columns]

In [20]:
# dropping meltan from the dataframe
df2 = df2[:-1]

## Exercise #3 - Scraping Individual Pages


### Step 3.1: Scrape the main image for Bulbasaur in a general way that could be applied to other pokemon pages by searching for the relevant tag and extracting the image URL.

In [21]:
def image_scrape(imgsrc):
    url = 'https://pokemondb.net/pokedex/' + imgsrc
    content = requests.get(url).content
    soup1 = BeautifulSoup(content,'html.parser')
    image_tags = soup1.find_all('img', {'src':re.compile('.jpg')})
    for image_tag in image_tags:
        print("THE IMAGE URL IS : {}" .format(image_tag.get('src')))

In [22]:
image_scrape('bulbasaur')

THE IMAGE URL IS : https://img.pokemondb.net/artwork/bulbasaur.jpg


### Display the image in your Jupyter notebook using code

In [23]:
Image (url = "https://img.pokemondb.net/artwork/bulbasaur.jpg")

### Step 3.2: Extract the location table

In [24]:
tables = pd.read_html(requests.get("https://pokemondb.net/pokedex/bulbasaur",headers={'User-agent': 'Mozilla/5.0'}).text)[-2]

In [25]:
display(HTML(tables.to_html()))

### Step 3.3 : Transpose the DataFrame such that each column is a video game and each row/cell is the location where you find Bulbasaur in that game

In [26]:
# taking tranpose using .T
tables_final = tables.T

In [27]:
for i in range(0,20):
    if i in tables_final.columns:
        tables_final.rename(columns={i:tables_final[i][0]}, inplace=True)
    else:
        break
tables_final_new = tables_final.iloc[1:]
display(HTML(tables_final_new.to_html(index = False)))

### Step 3.4 : Extract the location table and transpose it for all the pokemon in the sample

In [28]:
import time
loca_list = []
for links in df2['URL']:
    url = "https://pokemondb.net" + links
    tables_new = pd.read_html(requests.get(url, headers={'User-agent': 'Mozilla/5.0'}).text)[-2]
    tab = pd.DataFrame(tables_new)
    tables_final = tab.T
    for i in range(0,20):
        if i in tables_final.columns:
            tables_final.rename(columns={i:tables_final[i][0]}, inplace=True)
        else:
            break
    tables_final_new = tables_final.iloc[1:]
    xy = []
    if 'XY' in tables_final_new.columns:
        xy.append(url)
        xy.append(tables_final_new['XY'][1])
    else:
        continue
    print_df = pd.DataFrame(xy)
    loca_list.append(xy)
    time.sleep(0.1)

In [29]:
rw = pd.DataFrame(loca_list)
rw = rw.rename(columns={0: "URL", 1: "Location"})
display(HTML(rw.to_html(index = False)))

## Exercise #4

### Step 4.1 : Use the full sample of pokemon from the pokedex DataFrame. Create a table that shows the average  attack and defense for each type. Each type should be a row. Average attack and defense should be columns. 

In [30]:
df2.groupby('Type').agg('mean')[['Att.', 'Def.']].rename(columns={'Att.' : 'Avg Attack', 'Def.' : 'Avg Defense'})

Avg Attack  Avg Defense
Type                                  
Bug             45.125000    53.125000
Bug Electric    77.000000    60.000000
Bug Fire        85.000000    55.000000
Bug Flying      61.666667    71.333333
Bug Ghost       90.000000    45.000000
...                   ...          ...
Water Grass     70.000000    70.000000
Water Ground    84.333333    72.666667
Water Poison    40.000000    35.000000
Water Psychic   75.000000   110.000000
Water Rock      78.000000   103.000000

[98 rows x 2 columns]

In [31]:
df2.groupby('Type').agg('mean')[['Att.', 'Def.']].max()

Att.    150.0
Def.    200.0
dtype: float64

### Which types have the highest average attack? 
### Dragon Electric has the highest average attack of 150
### Which types have the highest average defense? 
### Steel Ground has the highest average defense of 200

In [32]:
df2.groupby('Type').agg('mean')[['Att.', 'Def.']].min()

Att.    25.0
Def.    25.0
dtype: float64

### Which type has the lowest average attack?
### Psychic Fairy has the lowest average attack of 25
### Which type has the lowest average defense?
### Psychic Fairy has the lowest average defense of 25